node/src/types/value_or_chunk.rs

```diff
@@ -0,0 +1,172 @@
+    Value(V),
+    /// Represents a chunk of data with attached proof.
+    ChunkWithProof(ChunkWithProof),
+}
+
+/// Error returned when constructing an instance of [`ValueOrChunk`].
+#[derive(Debug, Error)]
+pub enum ChunkingError {
+    /// Merkle proof construction error.
+    MerkleConstruction(MerkleConstructionError),
+    /// Serialization error.
+    SerializationError(bytesrepr::Error),
+}
+
+impl Display for ChunkingError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        match self {
+            ChunkingError::MerkleConstruction(error) => f
+                .debug_tuple("MerkleConstructionError")
+                .field(&error)
+                .finish(),
+            ChunkingError::SerializationError(error) => {
+                f.debug_tuple("SerializationError").field(&error).finish()
+            }
+        }
+    }
+}
+
+impl<V> ValueOrChunk<V> {
+    /// Creates an instance of [`ValueOrChunk::Value`] if data size is less than or equal to
+    /// [`ChunkWithProof::CHUNK_SIZE_BYTES`] or a [`ValueOrChunk::ChunkWithProof`] if it is greater.
+    /// In the latter case it will return only the `chunk_index`-th chunk of the value's byte
+    /// representation.
+    pub fn new(data: V, chunk_index: u64) -> Result<Self, ChunkingError>
+    where
+        V: ToBytes,
+    {
+        let bytes = ToBytes::to_bytes(&data).map_err(ChunkingError::SerializationError)?;
+        // NOTE: Cannot accept the chunk size bytes as an argument without changing the
+        // IndexedMerkleProof. The chunk size there is hardcoded and will be used when
+        // determining the chunk.
+        if bytes.len() <= ChunkWithProof::CHUNK_SIZE_BYTES {
+            Ok(ValueOrChunk::Value(data))
+        } else {
+            let chunk_with_proof = ChunkWithProof::new(&bytes, chunk_index)
+                .map_err(ChunkingError::MerkleConstruction)?;
+            Ok(ValueOrChunk::ChunkWithProof(chunk_with_proof))
+        }
+    }
+}
+
+/// Represents an enum that can contain either a whole trie or a chunk of it.
+pub type TrieOrChunk = ValueOrChunk<TrieRaw>;
+
+impl<V: Debug> Display for ValueOrChunk<V> {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        match self {
+            ValueOrChunk::Value(data) => f.debug_tuple("Value").field(data).finish(),
+            ValueOrChunk::ChunkWithProof(chunk) => f
+                .debug_struct("ChunkWithProof")
+                .field("index", &chunk.proof().index())
+                .field("hash", &chunk.proof().root_hash())
+                .finish(),
+        }
+    }
+}
+
+/// Represents the ID of a `TrieOrChunk` - containing the index and the root hash.
+/// The root hash is the hash of the trie node as a whole.
+/// The index is the index of a chunk if the node's size is too large and requires chunking. For
+/// small nodes, it's always 0.
+#[derive(DataSize, Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq, Hash)]
+pub struct TrieOrChunkId(pub u64, pub Digest);
```

Comment:
Wouldn't such ID be exactly the same for `ValueOrChunk<V>` in general? If so, maybe it would make sense to rename it to `ValueOrChunkId`?

Reply:
I didn't want to optimize prematurely and it turned out to be a good decision - I've already done some work on chunking of `Vec<ExecutionResult>` and there we will have different IDs for pre-1.5 and post-1.5 b/c we can't identify the values just based on the data.